File: storage/db-core/src/main/java/com/coffee/gu/coupon/IssuedCouponRepositoryImpl.java
package com.coffee.gu.coupon;

import com.coffee.gu.enums.EntityStatus;
import com.querydsl.jpa.impl.JPAQueryFactory;
import org.jspecify.annotations.NonNull;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.coffee.gu.coupon.QCouponEntity.couponEntity;
import static com.coffee.gu.coupon.QIssuedCouponEntity.issuedCouponEntity;

@Repository
public class IssuedCouponRepositoryImpl implements IssuedCouponRepository{

    private final IssuedCouponJpaRepository issuedCouponJpaRepository;
    private final JPAQueryFactory jpaQueryFactory;

    public IssuedCouponRepositoryImpl(IssuedCouponJpaRepository issuedCouponJpaRepository, JPAQueryFactory jpaQueryFactory) {
        this.issuedCouponJpaRepository = issuedCouponJpaRepository;
        this.jpaQueryFactory = jpaQueryFactory;
    }

    @Override
    public IssuedCoupon save(IssuedCoupon issuedCoupon) {
        return issuedCouponJpaRepository.save(IssuedCouponEntity.from(issuedCoupon)).toModel(issuedCoupon.getCoupon());
    }

    @Override
    public List<IssuedCoupon> saveAll(List<IssuedCoupon> issuedCoupons) {
        List<IssuedCouponEntity> entities = issuedCoupons.stream()
                .map(IssuedCouponEntity::from)
                .toList();
        List<IssuedCouponEntity> savedEntities = issuedCouponJpaRepository.saveAll(entities);
        return savedEntities.stream()
                .map(entity -> entity.toModel(issuedCoupons.get(savedEntities.indexOf(entity)).getCoupon()))
                .toList();
    }

    @Override
    public List<IssuedCoupon> findAllByPrincipalKey(String principalKey) {
        List<IssuedCouponEntity> issuedCoupons = jpaQueryFactory
                .selectFrom(issuedCouponEntity)
                .where(
                        issuedCouponEntity.principalKey.eq(principalKey),
                        issuedCouponEntity.entityStatus.eq(EntityStatus.ACTIVE)
                )
                .fetch();
        return getIssuedCoupons(issuedCoupons);
    }

    @Override
    public boolean existsByPrincipalKeyAndCouponId(String principalKey, Long couponId) {
        return jpaQueryFactory.selectFrom(issuedCouponEntity)
                .where(
                        issuedCouponEntity.principalKey.eq(principalKey),
                        issuedCouponEntity.couponId.eq(couponId),
                        issuedCouponEntity.entityStatus.eq(EntityStatus.ACTIVE)
                )
                .fetchFirst() != null;
    }

    @Override
    public Optional<IssuedCoupon> findById(Long issuedCouponId) {
        IssuedCouponEntity issuedCoupon = jpaQueryFactory.selectFrom(issuedCouponEntity)
                .where(
                        issuedCouponEntity.id.eq(issuedCouponId),
                        issuedCouponEntity.entityStatus.eq(EntityStatus.ACTIVE)
                )
                .fetchFirst();
        if (issuedCoupon == null) return Optional.empty();
        CouponEntity coupon = jpaQueryFactory.
                selectFrom(couponEntity)
                .where(couponEntity.id.eq(issuedCoupon.getCouponId()))
                .fetchFirst();
        return Optional.of(issuedCoupon.toModel(coupon.toModel()));
    }

    @Override
    public List<IssuedCoupon> findAllByIdIn(List<Long> issuedCouponIds) {
        List<IssuedCouponEntity> issuedCoupons = jpaQueryFactory
                .selectFrom(issuedCouponEntity)
                .where(
                        issuedCouponEntity.id.in(issuedCouponIds),
                        issuedCouponEntity.entityStatus.eq(EntityStatus.ACTIVE)
                )
                .fetch();
        return getIssuedCoupons(issuedCoupons);
    }

    private List<IssuedCoupon> getIssuedCoupons(List<IssuedCouponEntity> issuedCoupons) {
        if (issuedCoupons.isEmpty()) return List.of();
        List<Long> couponIds = issuedCoupons.stream()
                .map(IssuedCouponEntity::getCouponId)
                .distinct()
                .toList();
        List<CouponEntity> coupons = jpaQueryFactory
                .selectFrom(couponEntity)
                .where(couponEntity.id.in(couponIds))
                .fetch();
        Map<Long, Coupon> couponMap = coupons.stream()
                .collect(Collectors.toMap(CouponEntity::getId, CouponEntity::toModel));
        return issuedCoupons.stream()
                .map(entity -> {
                    Coupon coupon = couponMap.get(entity.getCouponId());
                    return entity.toModel(coupon);
                })
                .toList();
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/coupon/IssuedCouponRepositoryImpl.kt
package com.coffee.gu.coupon

import com.coffee.gu.CoreException
import com.coffee.gu.ErrorType
import com.coffee.gu.coupon.QCouponEntity.couponEntity
import com.coffee.gu.coupon.QIssuedCouponEntity.issuedCouponEntity
import com.coffee.gu.enums.EntityStatus
import com.querydsl.jpa.impl.JPAQueryFactory
import jakarta.persistence.EntityNotFoundException
import org.springframework.stereotype.Repository
import java.util.Optional

@Repository
class IssuedCouponRepositoryImpl(
    private val issuedCouponJpaRepository: IssuedCouponJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory,
) : IssuedCouponRepository {

    override fun save(issuedCoupon: IssuedCoupon): IssuedCoupon {
        return issuedCouponJpaRepository.save(IssuedCouponEntity.from(issuedCoupon))
            .toModel(issuedCoupon.coupon)
    }

    override fun saveAll(issuedCoupons: List<IssuedCoupon>): List<IssuedCoupon> {
        val couponMap = issuedCoupons.map { it.coupon }.distinctBy { it.id }.associateBy { it.id }
        val entities = issuedCoupons.map { IssuedCouponEntity.from(it) }
        val savedEntities = issuedCouponJpaRepository.saveAll(entities)
        return savedEntities.mapNotNull { entity ->
            couponMap[entity.couponId]?.let { coupon ->
                entity.toModel(coupon)
            }
        }
    }

    override fun findAllByPrincipalKey(principalKey: String): List<IssuedCoupon> {
        val issuedCoupons = jpaQueryFactory
            .selectFrom(issuedCouponEntity)
            .where(
                issuedCouponEntity.principalKey.eq(principalKey),
                issuedCouponEntity.entityStatus.eq(EntityStatus.ACTIVE),
            )
            .fetch()
        return getIssuedCoupons(issuedCoupons)
    }

    override fun existsByPrincipalKeyAndCouponId(principalKey: String, couponId: Long): Boolean {
        return jpaQueryFactory.selectFrom(issuedCouponEntity)
            .where(
                issuedCouponEntity.principalKey.eq(principalKey),
                issuedCouponEntity.couponId.eq(couponId),
                issuedCouponEntity.entityStatus.eq(EntityStatus.ACTIVE),
            )
            .fetchFirst() != null
    }

    override fun findById(issuedCouponId: Long): IssuedCoupon {
        val issuedCoupon = issuedCouponJpaRepository.findById(issuedCouponId).orElseThrow { CoreException(ErrorType.NOT_FOUND_DATA) }
        val coupon = jpaQueryFactory
            .selectFrom(couponEntity)
            .where(couponEntity.id.eq(issuedCoupon.couponId))
            .fetchFirst()

        return issuedCoupon.toModel(coupon?.toModel() ?: throw CoreException(ErrorType.NOT_FOUND_DATA))
    }

    override fun findAllByIdIn(issuedCouponIds: List<Long>): List<IssuedCoupon> {
        val issuedCoupons = jpaQueryFactory
            .selectFrom(issuedCouponEntity)
            .where(
                issuedCouponEntity.id.`in`(issuedCouponIds),
                issuedCouponEntity.entityStatus.eq(EntityStatus.ACTIVE),
            )
            .fetch()
        return getIssuedCoupons(issuedCoupons)
    }

    private fun getIssuedCoupons(issuedCoupons: List<IssuedCouponEntity>): List<IssuedCoupon> {
        if (issuedCoupons.isEmpty()) return emptyList()
        val couponIds = issuedCoupons.map { it.couponId }.distinct()
        val coupons = jpaQueryFactory
            .selectFrom(couponEntity)
            .where(couponEntity.id.`in`(couponIds))
            .fetch()
        val couponMap = coupons.associateBy { it.id }
        return issuedCoupons.mapNotNull { entity ->
            val couponEntity = couponMap[entity.couponId] ?: return@mapNotNull null
            entity.toModel(couponEntity.toModel())
        }
    }
}
